addons/im_livechat/models/chatbot_script_step.py

```diff
@@ -0,0 +1,199 @@
+        """ Unlink the triggering answers that comes from the following steps after a sequence change """
+        if self.triggering_answer_ids:
+            update_command = [Command.unlink(answer.id) for answer in self.triggering_answer_ids
+                              if answer.step_id.sequence >= self.sequence]
+            if update_command:
+                self.write({'triggering_answer_ids': update_command})
+
+    def _get_parent(self):
+        """ Returns the first preceding step that matches either the triggering answers
+         or the possible answers the user can select """
+        self.ensure_one()
+
+        if not self.chatbot_id.ids:
+            return False
+
+        if not self.triggering_answer_ids:
+            return False
+        else:
+            domain = [
+                ('chatbot_id', 'in', self.chatbot_id.ids),  # Used ids instead of id to avoid testing on NewId
+                ('step_type', 'in', ['forward_operator', 'question_selection']),
+                ('sequence', '<', self.sequence),
+            ]
+            answer_domain = []
+            for answer in self.triggering_answer_ids:
+                answer_domain = expression.AND([answer_domain, [('triggering_answer_ids', 'in', answer.ids)]])
+            domain = expression.AND([domain, expression.OR([
+                answer_domain, [('answer_ids', 'in', self.triggering_answer_ids.ids)]])
+            ])
+        return self.search(domain, order='sequence desc', limit=1)
+
+    @api.depends('sequence', 'triggering_answer_ids')
+    def _compute_is_forward_operator_child(self):
+        for step in self:
+            parent = step._get_parent()
+            while parent:
+                if parent.step_type == 'forward_operator':
+                    break
+                parent = parent._get_parent()
+            step.is_forward_operator_child = parent and parent.step_type == 'forward_operator'
+
+    def _fetch_next_step(self, selected_answer_ids):
+        self.ensure_one()
+        domain = [('chatbot_id', '=', self.chatbot_id.id), ('sequence', '>', self.sequence)]
+        if selected_answer_ids:
+            domain = expression.AND([domain, [
+                '|',
+                ('triggering_answer_ids', '=', False),
+                ('triggering_answer_ids', 'in', selected_answer_ids.ids)]])
+        steps = self.env['im_livechat.chatbot.script_step'].sudo().search(domain)
+        for step in steps:
+            if not step.triggering_answer_ids:
+                return step
+            answers_by_step = defaultdict(list)
+            for answer in step.triggering_answer_ids:
+                answers_by_step[answer.step_id.id].append(answer)
+            if all(set(answers).issubset(set(selected_answer_ids))
+                   for (idx, answers) in answers_by_step.items() if len(answers) == 1):
+                return step
+        return self.env['im_livechat.chatbot.script_step']
+
+    def _filtered_welcome_steps(self):
```

Comment:
Seems this would be better defined on the ``chatbot.script`` model. You may define an interator that stops at first non text step, following the sequence. Defining it on the step level is weird in my opinion. In code you do ``chatbot.step_ids._filtered_welcome_steps`` so makes sense to have ``chatbot._filtered_welcome_steps`` .

PS: could also be called ``_get_welcome_steps`` in that case :)

Reply:
Indeed, this was a bit strange, Patrick moved it to the chatbot.script model and it makes more sense there.